shard: Fixed redundant checks done

rkothiya opened the pull request:
This patch fixes redundant checks done while calling
shard_modify_size_and_block_count.

Fixes: #1703

Change-Id: I735e532c78cbb181afa4b51480ad742ef4a75f77
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

Commit:
shard: Fixed redundant checks done

This patch fixes redundant checks done while calling
shard_modify_size_and_block_count.

Fixes: #1703

Change-Id: I735e532c78cbb181afa4b51480ad742ef4a75f77
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

deepshikhaaa commented:
/recheck smoke

schaffung reviewed (dismissed):
+1

xhernandez commented on xlators/features/shard/src/shard.c:
```diff
@@ -1590,8 +1590,7 @@ shard_lookup_cbk(call_frame_t *frame, void *cookie, xlator_t *this,
 
-    if (dict_get(xdata, GF_XATTR_SHARD_FILE_SIZE) &&
-        frame->root->pid != GF_CLIENT_PID_GSYNCD)
+    if (frame->root->pid != GF_CLIENT_PID_GSYNCD)
```
This is not exactly the same. If `dict_get()` fails here, it will silently ignore the issue and continue. This means that shard is not enabled for this file. However if you call `shard_modify_size_and_block_count()` when the xattr is not present, it will log a misleading error message.

We should check if the log message is really needed in some case and, if so, maybe passing a boolean to tell whether we want the log or not.

mykaul replied:
@xhernandez - but wasn't it supposed to be set on the lookup?
```
    if (frame->root->pid != GF_CLIENT_PID_GSYNCD) {
        ret = dict_set_uint64(local->xattr_req, GF_XATTR_SHARD_FILE_SIZE,
                              8 * 4);
        if (ret) {
            gf_msg(this->name, GF_LOG_WARNING, 0, SHARD_MSG_DICT_OP_FAILED,
                   "Failed to set dict value: key:%s for path %s.",
                   GF_XATTR_SHARD_FILE_SIZE, loc->path);
            goto err;
        }
    }
```

xhernandez replied:
In the lookup we add the `GF_XATTR_SHARD_FILE_SIZE` key to the dict so that the lookup fop in posix fills it with the existing value, but only if it's present (the value set here is irrelevant). If the requested xattr is not present in the backend, the cbk's xdata dict won't contain that key.

In this particular case, the file could have been created before enabling sharding, so the shard xattr won't be present and the file needs to be managed as if shard is not enabled. This case shouldn't be handled as an error, so we shouldn't log an error.

vh05 replied:
Agree with @xhernandez . We can either pass a boolean to ignore the msg or the return value of shard_modify_size_and_block_count() can be considered in the parent function to inform whether the dict_get() failure can be ignored and continue

mykaul replied:
Hmmm... What would be the block size in this case? Maybe we should have bailed out earlier with block size check?

(Partially related - I doubt it's a flow we regularly check, of enabling sharding on pre-existing files)

mykaul replied:
@VHariharmath-rh - I don't think it should be a boolean - if we already fetch the value in the caller function, we should pass the value along and skip the redundant fetch - that's the whole point of this code change.

xhernandez replied:
> Hmmm... What would be the block size in this case? Maybe we should have bailed out earlier with block size check?

If the block size is not defined, the file is handled as a regular non-sharded file (i.e. there are no shards in `.shard` and no special action needs to be done for requests on this file).

> 
> (Partially related - I doubt it's a flow we regularly check, of enabling sharding on pre-existing files)

No, I think we don't test it much. There are a couple of .t scripts that create files before enabling shard, but it doesn't seem they explicitly check the integrity of the file.

Commit:
shard: Fixed redundant checks done

Avoid printing log message when we do not know if key
GF_XATTR_SHARD_FILE_SIZE is present in dictionary.

Fixes: #1703

Change-Id: I1f6e30366e1fa2093cf6f5158e77227aa0d0f560
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/recheck smoke

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -548,19 +548,21 @@ shard_local_wipe(shard_local_t *local)
 
 int
-shard_modify_size_and_block_count(struct iatt *stbuf, dict_t *dict)
+shard_modify_size_and_block_count(struct iatt *stbuf, dict_t *dict,
```
nit: if you are already changing the signature of the function, make it static while at it.

rkothiya replied:
Done

mykaul commented on xlators/features/shard/src/shard.c:
```diff
@@ -548,19 +548,21 @@ shard_local_wipe(shard_local_t *local)
+        if (logerror)
+            gf_msg_callingfn(THIS->name, GF_LOG_ERROR, 0,
+                             SHARD_MSG_INTERNAL_XATTR_MISSING,
+                             "Failed to "
```
nit could be in one line, for easier grep

rkothiya replied:
Done

Commit:
shard: Fixed redundant checks done

As per the review comments made the function static.

Fixes: #1703

Change-Id: I54b756b32cd93076583c69b429192ee35e52155f
Signed-off-by: Rinku Kothiya <rkothiya@redhat.com>

rkothiya commented:
/run regression
